### PR TITLE
Homework - unique_ptr

### PR DESCRIPTION
**Pull request by mateusz-talma**

### CODE REVIEW

File: homework/unique_ptr/UniquePtr.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+template <typename T>
+UniquePtr<T>::UniquePtr(UniquePtr&& movedPtr) {
+    ptr_ = movedPtr.release();
+}
+
+template <typename T>
+UniquePtr<T>::~UniquePtr() {
+    if (ptr_) {
```
Comment (MaciejSalamonski) [inline homework/unique_ptr/UniquePtr.hpp:37]: You dont have to check if ptr is nullptr (in this case)

https://stackoverflow.com/questions/3116980/do-i-have-to-set-pointer-to-nullptr-in-destructor

**Review by MaciejSalamonski — APPROVED** [review]
Hi, nice code! Keep it up ;p

File: .github/workflows/uniquePtr.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+name: CMake CI
+
+on: [push, pull_request]
+
+jobs:
+  ctest:
```
Comment (ziobron) [inline .github/workflows/uniquePtr.yml:6]: add a valgrind / asan job as well to prove that you have no memory leaks.

File: homework/unique_ptr/UniquePtr-ut.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    UniquePtr<double> newDoublePtr = std::move(doublePtr);
+    ASSERT_EQ(*newDoublePtr, ConstValues::doubleVal);
+    ASSERT_EQ(doublePtr.get(), nullptr);
+    ASSERT_EQ(rawPtr, newDoublePtr.get());
+}
+
+TEST_F(UniquePtrTest, ShouldUseReset) {
+    doublePtr.reset(new double{ConstValues::newdoubleVal});
+    ASSERT_EQ(*doublePtr, ConstValues::newdoubleVal);
+}
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr-ut.cpp:48]: please also test a reset function without an argument
`doublePtr.reset();`

File: homework/unique_ptr/UniquePtr-ut.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    delete releasedPtr;
+}
+
+TEST_F(UniquePtrTest, ShouldCallMethodByArrowOperator) {
+    double a = 3.5;
+    double b = 2.0;
+    UniquePtr<Square> squarePtr = new Square(a, b);
+
+    ASSERT_EQ(squarePtr->ComputeArea(), a * b);
+}
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr-ut.cpp:64]: nice, please also test moving to the pointer that already holds another object.

File: homework/unique_ptr/UniquePtr.hpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        : ptr_(nullptr) {}
+    UniquePtr(T* ptr)
+        : ptr_(ptr) {}
+
+    UniquePtr(const UniquePtr&) = delete;
+    UniquePtr(UniquePtr&& movedPtr);
+
+    ~UniquePtr();
+
+    T& operator=(const T&) = delete;
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.hpp:18]: you are missing a move assignment operator

File: homework/unique_ptr/UniquePtr.hpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    T* operator->() const { return ptr_; }
+    T& operator*() const;
+
+    T* get() const { return ptr_; }
+    T* release();
+    void reset(T* newPtr);
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.hpp:24]:
```suggestion
    void reset(T* newPtr = nullptr);
```
the default argument is useful here

File: homework/unique_ptr/UniquePtr.hpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+template <typename T>
+T* UniquePtr<T>::release() {
+    T* res = nullptr;
+    std::swap(res, ptr_);
+    return res;
+}
+
+template <typename T>
+void UniquePtr<T>::reset(T* newPtr) {
+    if (ptr_) {
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.hpp:59]: if not needed. deleting nullptr is safe.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: homework/unique_ptr/UniquePtr.hpp (hunk 1 above)
Comment (WBigaj) [reply homework/unique_ptr/UniquePtr.hpp:37]: As @MaciejSalamonski told deleting nullptr is safe.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
{ "requirement_ids": [3784, 1119] }
36 XP za implementację unique_ptra (3784)
3 XP za punktualność

**Pull request closed without merging**